asdf/type_index.py

```diff
@@ -47,7 +47,6 @@ def __init__(self, version, index):
         self._class_by_subclass = {}
         self._types_with_dynamic_subclasses = {}
         self._extension_by_cls = {}
-        self._extensions_used = set()
```

Comment:
Keeping this state here is a bad idea because this object gets reused by all `AsdfFile` in the session, so files end up with extensions in their metadata that weren't actually used when serializing the file.

Reply:
Ah, this explains why I was seeing extensions not needed listed for a particular file. I had assumed this was a general bad feature, not that it was a path-dependent bug.